### PR TITLE
[FLINK-15585][table] Improve function identifier string in plan digest

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This improves the function identifier string for anonymous/inline functions. The new identifier has the following characteristics:

- If a function is not stateful, the identifier contains only the class name. No custom hash.
- The anonymous/inline function nature is visible via `*...*` similar to structured types.
- We use the pure class name to find classes easily (previously the information about inner classes was lost via `$` instead of `.`).

## Brief change log

- Improved `functionIdentifier()`
- Updates plans.

## Verifying this change

This change added tests and can be verified as follows: `UserFunctionHelperTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunction.java [hunk 1]
```diff
@@ -58,9 +60,13 @@
...
     /** Returns a unique, serialized representation for this function. */
     public final String functionIdentifier() {
+        final String className = getClass().getName();
+        if (isClassNameSerializable(this)) {
+            return className;
+        }
         final String md5 =
                 EncodingUtils.hex(EncodingUtils.md5(EncodingUtils.encodeObjectToString(this)));
-        return getClass().getName().replace('.', '$').concat("$").concat(md5);
+        return className.concat("$").concat(md5);
```
Comment (slinkydeveloper) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunction.java:69]: Does this logic works when the function is implemented as a scala `object`?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java [hunk 2]
```diff
@@ -143,34 +162,27 @@ private void testErrorMessage() {
...
         static TestSpec forInstance(UserDefinedFunction function) {
-            return new TestSpec(function);
+            return new TestSpec(null, function, null);
         }
 
-        static TestSpec forCatalogFunction(String className, FunctionLanguage language) {
-            return new TestSpec(new CatalogFunctionMock(className, language));
+        static TestSpec forCatalogFunction(
+                String className,
+                @SuppressWarnings("SameParameterValue") FunctionLanguage language) {
```
Comment (slinkydeveloper) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java:184]: What warning is this one?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java [hunk 3]
```diff
@@ -243,6 +244,36 @@ public static void prepareInstance(ReadableConfig config, UserDefinedFunction fu
...
+        while (!currentClass.equals(UserDefinedFunction.class)) {
+            for (Field field : currentClass.getDeclaredFields()) {
+                if (!Modifier.isTransient(field.getModifiers())
+                        && !Modifier.isStatic(field.getModifiers())) {
+                    // function seems to be stateful
+                    return false;
+                }
+            }
+            currentClass = currentClass.getSuperclass();
+        }
```
Comment (slinkydeveloper) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:267]: That sounds like a very magic way to find out whether a function is stateful or not. Don't we have a marker interface for that?

Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:267]: Even with a marker interface you need perform this check. User-defined function might come in various flavors. And the more interfaces, the harder it gets to get the job done.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java (hunk 2 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/functions/UserDefinedFunctionHelperTest.java:184]: You are right. I will simply always choose Java for now.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java (hunk 3 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:267]: It is, but on the other hand this is very magic and very hard as a behavior to predict for the users. I would rather have this kind of logic enabled behind a flag and default to "if you wanna implement a stateful function, mark it as such with this interface"

Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java:267]: I'm also against magic but in this case the magic is minimal. It only influences a plan digest and makes it more deterministic. Most users don't parameterize their functions anyway and/or register it under a name which makes them non-anonymous anyway.

**Comment by twalthr** [comment]
@slinkydeveloper I addressed the test feedback. Please have another look.

File: docs/content/docs/dev/table/functions/udfs.md [hunk 4]
```diff
@@ -240,7 +240,9 @@ An implementation class must extend from one of the available base classes (e.g.
 
 The class must be declared `public`, not `abstract`, and should be globally accessible. Thus, non-static inner or anonymous classes are not allowed.
 
-For storing a user-defined function in a persistent catalog, the class must have a default constructor and must be instantiable during runtime.
+For storing a user-defined function in a persistent catalog, the class must have a default constructor
+and must be instantiable during runtime. Anonymous functions in Table API can only be persisted if the
+function is not stateful (i.e. no non-transient fields).
```
Comment (slinkydeveloper) [inline docs/content/docs/dev/table/functions/udfs.md:245]: and non static.

**Review by slinkydeveloper — COMMENTED** [review]
Just one little comment and we're good

File: docs/content/docs/dev/table/functions/udfs.md (hunk 4 above)
Comment (twalthr) [reply docs/content/docs/dev/table/functions/udfs.md]: updated

**Review by slinkydeveloper — APPROVED** [review]

**Pull request closed without merging**